File: problems/CSES/Graph_Algorithms/message_route/message_route.py
import sys
from collections import deque
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    n, m = map(int, input().split())

    graph = [[] for _ in range(n + 1)]

    for _ in range(m):
        a, b = map(int, input().split())
        graph[a].append(b)
        graph[b].append(a)

    queue = deque([1])

    visited = [False] * (n + 1)
    visited[1] = True

    parent = [-1] * (n + 1)

    while queue:
        node = queue.popleft()

        for nei in graph[node]:
            if not visited[nei]:
                visited[nei] = True
                parent[nei] = node
                queue.append(nei)

    if not visited[n]:
        print("IMPOSSIBLE")
    else:
        path = []

        node = n
        while node != -1:
            path.append(node)
            node = parent[node]

        path.reverse()

        print(len(path))
        print(*path)

main()

File: problems/CSES/Graph_Algorithms/message_route/message_route.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    unordered_map<int, vector<int>> graph;
    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;

        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    queue<int> q;
    q.push(1);

    vector<bool> visited(n + 1, false);
    visited[1] = true;

    vector<int> parent(n + 1, -1);

    while (!q.empty()) {
        int node = q.front();
        q.pop();

        for (int nei : graph[node]) {
            if (!visited[nei]) {
                visited[nei] = true;
                parent[nei] = node;
                q.push(nei);
            }
        }
    }

    if (!visited[n]) {
        cout << "IMPOSSIBLE" << "\n";
    } else {
        vector<int> path;

        int node = n;
        while (node != -1) {
            path.push_back(node);
            node = parent[node];
        }

        reverse(path.begin(), path.end());

        cout << path.size() << '\n';

        for (auto num : path) cout << num << "\n";
    }

    return 0;
}
